### ISSUE TITLE
Deserialization fails depending on the order of deserialized objects with "Cannot construct instance (although at least one Creator exists)"

### ISSUE THREAD

**Issue by dinopraso** [FasterXML/jackson-databind#3355]
**Describe the bug**
Deserialization of a specific class becomes broken if another specific class is deserialized before it. All classes are immutable and depend on `@JsonCreator` and property-based constructors with all properties annotated correctly with `@JsonProperty`.

If a class `ContainerOne` is deserialized before a class `ContainerTwo` (see below for reference), deserialization will fail, but it you flip their order, it will work as expected.

It appears that some sort of caching is handled incorrectly making the second invocation ignore the property-based creator.

**Version information**
Reproduced on 2.12.3 and 2.13.0

**To Reproduce**
Here is the minimal setup I've managed to make to reproduce this.
```java
public class Common {
    private final String property;
    private final List<ContainerTwo> twos;

    @JsonCreator
    public Common(@JsonProperty("property") final String property,
                  @JsonProperty("twos") final List<ContainerTwo> twos) {
        this.property = property;
        this.twos = twos;
    }

    public String getProperty() {
        return property;
    }

    // @JsonIgnoreProperties("common") // This is commented out because it does not contribute to the reproduction, but would be present for such a setup.
    public List<ContainerTwo> getTwos() {
        return twos;
    }
}

public class ContainerOne {
    private final Common common;

    @JsonCreator
    public ContainerOne(@JsonProperty("common") final Common common) {
        this.common = common;
    }

    public Common getCommon() {
        return common;
    }
}

public class ContainerTwo {
    private final Common common;

    @JsonCreator
    public ContainerTwo(@JsonProperty("common") final Common common) {
        this.common = common;
    }

    @JsonIgnoreProperties("twos")
    public Common getCommon() {
        return common;
    }
}

public static void main(final String[] args) throws Exception {
    final String oneJson = "{ \"common\": { \"property\": \"valueOne\" } }";
    final String twoJson = "{ \"common\": { \"property\": \"valueTwo\" } }";

    final ObjectMapper objectMapper = new ObjectMapper();

    final ContainerOne one = objectMapper.readValue(oneJson, ContainerOne.class);
    final ContainerTwo two = objectMapper.readValue(twoJson, ContainerTwo.class); // This fails, unless invoked before the line directly above.
}
```
I have found two ways to stop this from happening, both of which are not really an option for me at this point:
1. Remove `@JsonIgnoreProperties("twos")` from the getter in `ContainerTwo`
2. Make `Common` mutable, with setters

**Expected behavior**
Since this behavior is specific to the execution order and breaks seemingly unexpectedly, it appears to be a bug.
Expected behavior should be that both objects are deserialized (like they are individually) regardless of their invocation order.

**Comment by cowtowncoder** [comment]
It does sound like a bug, and my first guess would be it might be related to `@JsonIgnoreProperties` handling.

**Comment by dinopraso** [comment]
After some more fiddling around, I've noticed that that `ContainerOne` is not necessary for the reproduction of the issue. Here is an even simpler example (night help in locating the issue easier)
```java
public class Common {
    private final String property;
    private final List<Container> containers;

    @JsonCreator
    public Common(@JsonProperty("property") final String property,
                  @JsonProperty("containers") final List<Container> containers) {
        this.property = property;
        this.containers = containers;
    }

    public String getProperty() {
        return property;
    }

    public List<Container> getContainers() {
        return containers;
    }
}

public class Container {
    private final Common common;

    @JsonCreator
    public Container(@JsonProperty("common") final Common common) {
        this.common = common;
    }

    @JsonIgnoreProperties("containers")
    public Common getCommon() {
        return common;
    }
}

public static void main(final String[] args) throws Exception {
    final String objectJson = "{ \"property\": \"valueOne\" }";
    final String containersJson = "{ \"common\": { \"property\": \"valueTwo\" } }";

    final ObjectMapper objectMapper = new ObjectMapper();

    final Common object = objectMapper.readValue(objectJson, Common.class);
    final Container container = objectMapper.readValue(containersJson, Container.class); // Fails
}
```

Also, it seems to be only an issue with the property-based creator. Adding any other creator (setters or POJOBuilder) makes the issue go away, though both are not ideal workarounds.

**Comment by dinopraso** [comment]
Can we get any updates on this? It's fairly important for us to have this fixed soon, and we would like to avoid having to write temporary workarounds in out application.

**Comment by cowtowncoder** [comment]
@dinopraso Unfortunately I don't have time to work on this at this point in time. I hope sometimes else can help, or you figure it out.

**Comment by yawkat** [comment]
Yes this is also the behavior for other annotations, the best example being `@JsonProperty`:

```java
class Test {
    private String foo;

    @JsonProperty("bar")
    public String getFoo() {
        return foo;
    }

    public void setFoo(String foo) {
        this.foo = foo;
    }
}
```

The property will be called `bar` for both serialization and deserialization.

You can essentially think of annotations as applying to the *property* (getter + setter), not the individual accessor, though there are some special rules when it comes to conflicting annotations.

**Comment by dinopraso** [comment]
Though the issue still remains to be fixed, since the fact that a set of properties was ignored on one class shouldn't affect de deserialization of another

**Comment by cowtowncoder** [comment]
The behavior about "unifying annotations" (combining) -- that annotations in one accessor are indeed applied to the "whole" logical property, regardless of which accessor had it -- is indeed 100% intended and working as defined.
The reason for this is to reduce the need to apply multiple duplicate annotations for things like renaming (as illustrated by @yawkat's example).
The process of unification is applied for all Jackson annotations without explicit rules in trying to determine separate semantics for separate annotations.
But even if handling were separated, I would argue that `@JsonIgnoreProperties` should apply regardless of setter/getter used to denote it.

Having said that it sounds like there is a separate issue which is likely due to the way that `@JsonIgnoreProperties` is applied via reference. It sounds likely that the "modified" instance (with ignoral) gets cached and incorrectly used for non-annotated case.
I think there is another existing issue filed for recursive use case that is probably related.

**Comment by pparam5241** [comment]
In order to fix this issue, You may create your JsonDeserializer which will help you to convert this.

I've used below snippet in order to fix my issue which works fine.

```
public class JsonStringDeserializer extends JsonDeserializer<Value> {

    @Override
    public Value deserialize(JsonParser jsonParser, DeserializationContext context) throws IOException {
        ObjectMapper mapper = (ObjectMapper) jsonParser.getCodec();
        String jsonString = jsonParser.getText();
        return mapper.readValue(jsonString, Value.class);
    }
}
```
Here, Value is my class.

**Comment by cowtowncoder** [comment]
If anyone has time, would be good to see if this still occurs with 2.18.2 -- good chance it's been fixed.

**Comment by cowtowncoder** [comment]
I think this might have same root cause as #1622, basically.

**Comment by JacksonJang** [comment]
@cowtowncoder 
That's correct.
I've already fixed the issue.
I'll clean up the code, add tests, and then submit a pull request.

**Comment by cowtowncoder** [comment]
Good luck @JacksonJang -- this may be somewhat challenging (#1622 definitely is) but it'd be great to solve the issue.
